File: code/go/ch21/src/luago/state/api_push.go
package state

import "fmt"
import . "luago/api"

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_pushnil
func (self *luaState) PushNil() {
	self.stack.push(nil)
}

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_pushboolean
func (self *luaState) PushBoolean(b bool) {
	self.stack.push(b)
}

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_pushinteger
func (self *luaState) PushInteger(n int64) {
	self.stack.push(n)
}

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_pushnumber
func (self *luaState) PushNumber(n float64) {
	self.stack.push(n)
}

// [-0, +1, m]
// http://www.lua.org/manual/5.3/manual.html#lua_pushstring
func (self *luaState) PushString(s string) {
	self.stack.push(s)
}

// [-0, +1, e]
// http://www.lua.org/manual/5.3/manual.html#lua_pushfstring
func (self *luaState) PushFString(fmtStr string, a ...interface{}) {
	str := fmt.Sprintf(fmtStr, a...)
	self.stack.push(str)
}

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_pushcfunction
func (self *luaState) PushGoFunction(f GoFunction) {
	self.stack.push(newGoClosure(f, 0))
}

// [-n, +1, m]
// http://www.lua.org/manual/5.3/manual.html#lua_pushcclosure
func (self *luaState) PushGoClosure(f GoFunction, n int) {
	closure := newGoClosure(f, n)
	for i := n; i > 0; i-- {
		val := self.stack.pop()
		closure.upvals[i-1] = &upvalue{&val}
	}
	self.stack.push(closure)
}

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_pushglobaltable
func (self *luaState) PushGlobalTable() {
	global := self.registry.get(LUA_RIDX_GLOBALS)
	self.stack.push(global)
}

// [-0, +1, –]
// http://www.lua.org/manual/5.3/manual.html#lua_pushthread
func (self *luaState) PushThread() bool {
	self.stack.push(self)
	return self.isMainThread()
}

File: code/csharp/ch13/luavm/state/ApiPush.cs
using luavm.api;

namespace luavm.state
{
    public partial class LuaState
    {
        public void PushNil()
        {
            _stack.Push(null);
        }

        public void PushBoolean(bool b)
        {
            _stack.Push(b);
        }

        public void PushInteger(long n)
        {
            _stack.Push(n);
        }

        public void PushNumber(double n)
        {
            _stack.Push(n);
        }

        public void PushString(string s)
        {
            _stack.Push(s);
        }

        public void PushCsharpFunction(CsharpFunction f)
        {
            _stack.Push(Closure.NewCsharpClosure(f, 0));
        }

        public void PushGlobalTable()
        {
            var global = Registry.Get(Constant.LUA_RIDX_GLOBALS);
            _stack.Push(global);
        }

        public void PushCsharpClosure(CsharpFunction f, int n)
        {
            var closure = Closure.NewCsharpClosure(f, n);
            for (var i = n; i > 0; i--)
            {
                var val = _stack.Pop();
                closure.Upvals[n - 1] = new Upvalue(val);
            }

            _stack.Push(closure);
        }
    }
}